src/ray/core_worker/transport/concurrency_group_manager.h

```diff
@@ -0,0 +1,113 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+namespace ray {
+
+namespace core {
+
+/// A manager that manages a set of concurrency group executors, which will perform
+/// the methods defined in one concurrency group.
+template <typename ConcurrencyGroupExecutorType>
+class ConcurrencyGroupManager final {
+ public:
+  explicit ConcurrencyGroupManager(
+      const std::vector<ConcurrencyGroup> &concurrency_groups = {},
+      const int32_t max_concurrency_for_default_concurrency_group = 1) {
+    for (auto &group : concurrency_groups) {
+      const auto name = group.name;
+      const auto max_concurrency = group.max_concurrency;
+      auto executor = std::make_shared<ConcurrencyGroupExecutorType>(max_concurrency);
+      auto &fds = group.function_descriptors;
+      for (auto fd : fds) {
+        functions_to_executor_index_[fd->ToString()] = executor;
+      }
+      name_to_executor_index_[name] = executor;
+    }
+
+    // If max concurrency of default group is 1 and there is no other concurrency group of
+    // this actor, the tasks of default group will be performed in main thread instead of
+    // any executor pool, otherwise tasks in any concurrency group should be performed in
+    // the thread pools instead of main thread.
+    if (ConcurrencyGroupExecutorType::NeedDefaultExecutor(
+            max_concurrency_for_default_concurrency_group) ||
+        !concurrency_groups.empty()) {
+      defatult_executor_ = std::make_shared<ConcurrencyGroupExecutorType>(
+          max_concurrency_for_default_concurrency_group);
+    }
+  }
+
+  /// Get the corresponding concurrency group executor by the give concurrency group or
+  /// function descriptor.
+  ///
+  /// Return the corresponding executor of the concurrency group
```

Comment:
Hmm the behavior is a little weird... But since it is refactoring we probably don't need to touch it..

Can you also follow the docstring convention we have here? \param and \return